strawberryfields/engine.py

```diff
@@ -647,23 +645,19 @@ def run(
             strawberryfields.api.Result, None: the job result if successful, and ``None`` otherwise
 
         Raises:
-            requests.exceptions.RequestException: if there was an issue sending
-                a request to the Xanadu Cloud
+            requests.exceptions.RequestException: if there was an issue fetching
+                the device specifications from the Xanadu Cloud
+            FailedJobError: if the remote job fails on the server side ("cancelled" or "failed")
         """
         job = self.run_async(
             program, compile_options=compile_options, recompile=recompile, **kwargs
         )
         try:
-            while True:
-                job.refresh()
-                if job.status == "complete":
-                    self.log.info("The remote job %s has been completed.", job.id)
-                    return job.result
-
+            while not job.finished:
```

Comment:
Okay, I admit that this part is slightly tricky. In an ideal world, we could just use
```python
job.wait()
```
but this would be problematic for long-running jobs that use a Connection without a refresh token (since the access token could expire before the job finishes). So, just to be safe, we can change the loop to
```python
while True:
    # Refresh the Connection
    job._connection = self.connection
    # Clear the Job cache
    job.clear()

    if job.finished:
        break
    time.sleep(self.POLLING_INTERVAL_SECONDS)
```
or
```python
while True:
    # Refresh the Connection and clear the Job cache
    job = xcc.Job(id_ = job.id, connection=self.connection)

    if job.finished:
        break
    time.sleep(self.POLLING_INTERVAL_SECONDS)
```
or even
```python
while not xcc.Job(job.id, self.connection).finished:
    time.sleep(self.POLLING_INTERVAL_SECONDS)
```
Note that the return value of virtually all `xcc.Job` properties are cached so we need to execute
```python
job.clear()
```
to fetch the updated status of a job.

Reply:
Thanks @Mandrenkov. I went with the first suggestion here, and we can revisit this once support for dynamically refreshing config info exist in the `xcc.Connection` class.